### DISCUSSION TITLE
Simple example of C/C++ program parsing Ruby

### DISCUSSION THREAD

**Discussion by calicoday** [tree-sitter/tree-sitter#1221] [Q&A]
The 'Using Parsers' section of the tree-sitter docs shows an example of a simple C program that uses the Tree-sitter JSON parser and the commands to compile and run it. An equivalent example using the Ruby parser would be useful, since there's the extra matter of the C++ external scanner. 

It's been donkey's years since I've done C/C++ but I did want to try the parser plain before layering in more wraps and, much helped by the work in tree-sitter issues #543 and #548, I was able to make a near-as-poss C++ example, test-ruby-parser.cc, by changing only:

- refer to Ruby rather than JSON, so s/json/ruby/g and s/JSON/Ruby/g
- include c++ headers not c, eg <cassert> not <assert.h>
- extern "C" the call for a language pointer
- test data needs to be ruby, so "[1, nil]" not "[1, null]"
- ast nodes types will be ruby, so "program" and "integer" not "document" and "number" in the asserts
- compile with clang++

So, test-ruby-parser.cc is:
```ruby
// Filename - test-ruby-parser.cc

#include <cassert> // was C headers
#include <cstring>
#include <cstdio>
#include <tree_sitter/api.h>

#ifdef __cplusplus
extern "C" { // was not extern "C"
#endif
// Declare the `tree_sitter_ruby` function, which is
// implemented by the `tree-sitter-ruby` library.
TSLanguage *tree_sitter_ruby();
#ifdef __cplusplus
}
#endif

int main() {
  // Create a parser.
  TSParser *parser = ts_parser_new();

  // Set the parser's language (Ruby in this case).
  ts_parser_set_language(parser, tree_sitter_ruby());

  // Build a syntax tree based on source code stored in a string.
  const char *source_code = "[1, nil]"; // was "[1, null]"
  TSTree *tree = ts_parser_parse_string(
    parser,
    NULL,
    source_code,
    strlen(source_code)
  );

  // Get the root node of the syntax tree.
  TSNode root_node = ts_tree_root_node(tree);

  // Get some child nodes.
  TSNode array_node = ts_node_named_child(root_node, 0);
  TSNode number_node = ts_node_named_child(array_node, 0);

  // Check that the nodes have the expected types.
  assert(strcmp(ts_node_type(root_node), "program") == 0); // was "document"
  assert(strcmp(ts_node_type(array_node), "array") == 0);
  assert(strcmp(ts_node_type(number_node), "integer") == 0); // was "number"

  // Check that the nodes have the expected child counts.
  assert(ts_node_child_count(root_node) == 1);
  assert(ts_node_child_count(array_node) == 5);
  assert(ts_node_named_child_count(array_node) == 2);
  assert(ts_node_child_count(number_node) == 0);

  // Print the syntax tree as an S-expression.
  char *string = ts_node_string(root_node);
  printf("Syntax tree: %s\n", string);

  // Free all of the heap-allocated memory.
  free(string);
  ts_tree_delete(tree);
  ts_parser_delete(parser);
  return 0;
}
```
and compile (ignoring `clang: warning: treating 'c' input as 'c++' when in C++ mode, this behavior is deprecated [-Wdeprecated]` because of parser.c) and run:
```
clang++ \
	-I ln-tree-sitter/lib/include \
	test-ruby-parser.cc \
	ln-ruby-src/parser.c \
	ln-ruby-src/scanner.cc \
	ln-tree-sitter/libtree-sitter.a \
	-o test-ruby-parser

./test-ruby-parser
```
which gives results:
```
Syntax tree: (program (array (integer) (nil)))
```

What I still can't figure is how one might make a C example with C parser.c and C++ scanner.c (I couldn't get it to link). If that's a sane or even possible thing to do? Anyway, this works for my immediate purposes and I thought I'd mention it, in case it's helpful for someone else.

**Comment by xiaolitongxue666** [comment]
I also want to find a example for a C example with C parser.c, why tree-sitter-c use js?

**Comment by ahlinc** [comment]
@xiaolitongxue666 JS is only used as a DSL on a one of parser generation phases.

**Reply by xiaolitongxue666** [reply]
Got it , maybe my build process not right , thanks ~
